Libraries/acdriver/Sources/CompileActionImageSet.cpp

```diff
@@ -0,0 +1,498 @@
+        case PNG_COLOR_TYPE_GRAY_ALPHA:
+            channels = 2;
+            break;
+        case PNG_COLOR_TYPE_PALETTE:
+        case PNG_COLOR_TYPE_RGB:
+            /* These are transformed to RGBA */
+        case PNG_COLOR_TYPE_RGB_ALPHA:
+            channels = 4;
+            break;
+        default:
+            png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, NULL);
+            fclose(fp);
+            result->normal(
+                Result::Severity::Error,
+                "Unhandled PNG color type");
+            return false;
+    }
+
+    row_bytes = png_get_rowbytes (png_struct_ptr, info_struct_ptr);
+    if (row_bytes != (width * (bit_depth/8) * channels)) {
+        png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, NULL);
+        result->normal(
+            Result::Severity::Error,
+            "Unable to transform PNG pixel data");
+        return false;
+    }
+
+    row_pointers = (png_byte **) malloc (height * sizeof (png_bytep));
+    if (row_pointers == NULL)
+    {
+        png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, NULL);
+        result->normal(
+            Result::Severity::Error,
+            "Could not allocate memory");
+        return false;
+    }
+
+    pixels = std::vector<unsigned char>(width * height * channels);
+    unsigned char *bytes = static_cast<unsigned char *>(pixels.data());
+    for (int row = 0; row < height; row++) {
+        row_pointers[row] = bytes + (row * row_bytes);
+    }
+    png_read_image(png_struct_ptr, row_pointers);
+
+    /* clean up */
+    png_read_end (png_struct_ptr, info_struct_ptr);
+    png_destroy_read_struct (&png_struct_ptr, &info_struct_ptr, (png_infopp) NULL);
+    free(row_pointers);
+    fclose(fp);
+
+    if (width_out) {
+        *width_out = width;
+    }
+    if (height_out) {
+        *height_out = height;
+    }
+    if (channels_out) {
+        *channels_out = channels;
+    }
+    return true;
+}
+
+static uint16_t
+GenerateIdentifier(void) {
+    static uint16_t last = 0;
+    last = last + 1;
+    return last;
+}
+
+bool
+CompileAsset(
+    std::shared_ptr<xcassets::Asset::Catalog> const &catalog,
+    std::string ns,
+    Options const &options,
+    CompileOutput *compileOutput,
+    Result *result,
+    std::shared_ptr<xcassets::Asset::Asset> const &parent,
+    xcassets::Asset::ImageSet::Image const &image)
+{
+    static std::map<std::string, uint16_t> idMap = {};
+    std::string name = FSUtil::GetBaseNameWithoutExtension(parent->path());
+
+    /* Skip any entry that is not attached to a file */
+    if (!image.fileName()) {
+        return true;
+    }
+
+    std::string filename = parent->path() + std::string("/") + *image.fileName();
+
+    /* An image without an idiom is considered unassigned */
+    if (!image.idiom()) {
+        return false;
+    }
+
+    /* Add namespace to name */
+    if (ns.size() > 0) {
+        name = ns + std::string("/") + name;        
```

Comment:
Instead of passing around and building up the namespaced asset name, you can use `asset->name().string()`. It's already built up while loading into an `xcassets::FullyQualifiedName`.